nemo/collections/asr/models/rnnt_models.py

```diff
@@ -0,0 +1,567 @@
+            with open_dict(self.cfg.joint):
+                self.cfg.joint = new_joint_config
+
+            with open_dict(self.cfg.decoder):
+                self.cfg.decoder = new_decoder_config
+
+            with open_dict(self.cfg.decoding):
+                self.cfg.decoding = decoding_cfg
+
+            logging.info(f"Changed decoder to output to {self.joint.vocabulary} vocabulary.")
+
+    def _setup_dataloader_from_config(self, config: Optional[Dict]):
+        if 'augmentor' in config:
+            augmentor = process_augmentations(config['augmentor'])
+        else:
+            augmentor = None
+
+        shuffle = config['shuffle']
+        device = 'gpu' if torch.cuda.is_available() else 'cpu'
+        if config.get('use_dali', False):
+            device_id = self.local_rank if device == 'gpu' else None
+            dataset = audio_to_text_dataset.get_dali_char_dataset(
+                config=config,
+                shuffle=shuffle,
+                device_id=device_id,
+                global_rank=self.global_rank,
+                world_size=self.world_size,
+                preprocessor_cfg=self._cfg.preprocessor,
+            )
+            return dataset
+
+        # Instantiate tarred dataset loader or normal dataset loader
+        if config.get('is_tarred', False):
+            if ('tarred_audio_filepaths' in config and config['tarred_audio_filepaths'] is None) or (
+                'manifest_filepath' in config and config['manifest_filepath'] is None
+            ):
+                logging.warning(
+                    "Could not load dataset as `manifest_filepath` was None or "
+                    f"`tarred_audio_filepaths` is None. Provided config : {config}"
+                )
+                return None
+
+            shuffle_n = config.get('shuffle_n', 4 * config['batch_size'])
+            dataset = audio_to_text_dataset.get_tarred_char_dataset(
+                config=config,
+                shuffle_n=shuffle_n,
+                global_rank=self.global_rank,
+                world_size=self.world_size,
+                augmentor=augmentor,
+            )
+            shuffle = False
+        else:
+            if 'manifest_filepath' in config and config['manifest_filepath'] is None:
+                logging.warning(f"Could not load dataset as `manifest_filepath` was None. Provided config : {config}")
+                return None
+
+            dataset = audio_to_text_dataset.get_char_dataset(config=config, augmentor=augmentor)
+
+        return torch.utils.data.DataLoader(
+            dataset=dataset,
+            batch_size=config['batch_size'],
+            collate_fn=dataset.collate_fn,
+            drop_last=config.get('drop_last', False),
+            shuffle=shuffle,
+            num_workers=config.get('num_workers', 0),
+            pin_memory=config.get('pin_memory', False),
+        )
+
+    def setup_training_data(self, train_data_config: Optional[Union[DictConfig, Dict]]):
```

Comment:
Hmm... methods like this make me increasingly support having a unified base ASR class, so that if one thing changes we don't have to go and fix a bunch of files.
Not really something specific to this PR, just something to keep in mind as we move forward

Reply:
Agreed entirely. I initially attempted that unification in earlier attempts, but the scale of the PR is daunting without additional core refactors. 

At least the BPE mechanism is abstracted away inside a Mixin (so CTC BPE and RNNT BPE can share the tokenization logic), and now datasets are (mostly) shared between the two domains. There's still a lot of stuff that's duplicated (mainly transcription, train_test_val setups, handling of multi_* datasets etc).

Needs a full scale refactor for sharing more components